File: python/deserialize.py
#!/usr/bin/env python3
"""Deserializes object from STDIN or file.

   SPDX-License-Identifier: GPL-3.0-or-later

   This file is part of the deserialization demos:
   https://github.com/PeterMosmans/deserialization-demos
 """

import pickle
import sys
from typing import Optional


import typer
import yaml


def read_file(filename: str):
    """Return contents of a file."""
    try:
        with open(filename, "rb") as file:
            contents = file.read()
        return contents
    except Exception as exception:
        print(f"Could not read {filename}: {exception}")
        sys.exit(-1)


def read_stdin():
    """Read stdin."""
    return sys.stdin.buffer.read()


def main(filename: Optional[str] = typer.Argument(None), format: str = "native"):
    """Deserialize a file to an object."""
    if filename:
        contents = read_file(filename)
    else:
        contents = read_stdin()
    try:
        if format == "yaml":
            deserialized = yaml.unsafe_load(contents)
        else:
            deserialized = pickle.loads(contents)
    except Exception as exception:
        print(f"Could not properly deserialize file: {exception}")
        sys.exit(-1)
    print(f"Object deserialized: {type(deserialized)}")


if __name__ == "__main__":
    typer.run(main)

File: java/Deserialize.java
/* Deserializes object from STDIN or file.
   The class definition is read from MyObject.

   SPDX-License-Identifier: GPL-3.0-or-later

   This file is part of the deserialization demos:
   https://github.com/PeterMosmans/deserialization-demos
 */

package net.gofwd;

import java.io.*;

public class Deserialize {

  // Serialize an object to a file
  public static void main(String[] args) {
    MyObject object = null;
    try {
      if (args.length == 0) {
        ObjectInputStream in = new ObjectInputStream(System.in);
        object = (MyObject) in.readObject();
        in.close();
      } else {
        String filename = args[0];
        FileInputStream file = new FileInputStream(filename);
        ObjectInputStream in = new ObjectInputStream(file);
        object = (MyObject) in.readObject();
        in.close();
        file.close();
      }
      System.out.println("Object deserialized: admin is " + object.admin);
    } catch (IOException ex) {
      System.out.println("Could not read: " + ex);
    } catch (ClassNotFoundException ex) {
      System.out.println("Class not found: " + ex);
    } catch (NullPointerException ex) {
      System.out.println("Could not deserialize object: " + ex);
    }
  }
}
